tools/dds/dds-sniffer/dds-sniffer.cpp

```diff
@@ -0,0 +1,302 @@
+        return EXIT_FAILURE;
+    }
+
+    return EXIT_SUCCESS;
+}
+catch( const TCLAP::ExitException & )
+{
+    std::cerr << "Undefined exception while parsing command line arguments" << std::endl;
+    return EXIT_FAILURE;
+}
+catch( const TCLAP::ArgException & e )
+{
+    std::cerr << e.what() << std::endl;
+    return EXIT_FAILURE;
+}
+catch( const std::exception & e )
+{
+    std::cerr << e.what() << std::endl;
+    return EXIT_FAILURE;
+}
+
+Sniffer::Sniffer()
+    : _participant( nullptr )
+    , _subscriber( nullptr )
+    , _topics()
+    , _readers()
+    , _datas()
+    , _subscriberAttributes()
+    , _readerQoS()
+{
+}
+
+Sniffer::~Sniffer()
+{
+    _running = false;
+
+    for( const auto & it : _topics )
+    {
+        if( _subscriber != nullptr )
+        {
+            _subscriber->delete_datareader( it.first );
+        }
+        _participant->delete_topic( it.second );
+    }
+    if( _subscriber != nullptr )
+    {
+        if( _participant != nullptr )
+        {
+            _participant->delete_subscriber( _subscriber );
+        }
+    }
+    if( _participant != nullptr )
+    {
+        eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->delete_participant( _participant );
+    }
+
+    _datas.clear();
+    _topics.clear();
+    _readers.clear();
+}
+
+bool Sniffer::init( eprosima::fastdds::dds::DomainId_t domain )
+{
+    eprosima::fastdds::dds::DomainParticipantQos participantQoS;
+    participantQoS.name( "dds-sniffer" );
+    _participant = eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->create_participant( domain,
+                                                                                                         participantQoS,
+                                                                                                         this );
+
+    std::cout << "Creating a sniffer for domain " << domain << std::endl;
+
+    return _participant != nullptr;
+}
+
+void Sniffer::run( uint32_t seconds )
+{
+    _running = true;
+
+    if( seconds == 0 )
+    {
+        std::cin.ignore( std::numeric_limits< std::streamsize >::max() );
+    }
+    else
+    {
+        std::this_thread::sleep_for( std::chrono::seconds( seconds ) );
+    }
+}
+
+void Sniffer::on_data_available( eprosima::fastdds::dds::DataReader * reader )
+{
+    if( _running )
+    {
+        auto dataIter = _datas.find( reader );
+
+        if( dataIter != _datas.end() )
+        {
+            eprosima::fastrtps::types::DynamicData_ptr data = dataIter->second;
+            eprosima::fastdds::dds::SampleInfo info;
+            if( reader->take_next_sample( data.get(), &info ) == ReturnCode_t::RETCODE_OK )
+            {
+                if( info.instance_state == eprosima::fastdds::dds::ALIVE_INSTANCE_STATE )
+                {
+                    eprosima::fastrtps::types::DynamicType_ptr type = _readers[reader];
+                    std::cout << "Received data of type " << type->get_name() << std::endl;
```

Comment:
Is the type indicative of the topic? Can you give an example?

Reply:
The type is the underlying data structure (usually defined in the IDL) of the topic.
For example in the Shaped Demo the topics are Square/Circle/Triangle of type Shape.